### PR TITLE
Create a Calendar Calculator and setup test suite

### PR DESCRIPTION
**Pull request by tonicospinelli**
It is responsible for calculating and manipulating Dates in the financial process, is a vital feature, and requires significant coverage.

The first test case is to create an understanding of how it works.

Changes in this pull request:

- add unit tests for the Support/Navigation class
- add a calendar calculator with periodicity strategies to calculate the next date
- avoid monthly (also the variations) and yearly to not overflow the next date
- setup test suites on the SonarCloud workflow and upload their code coverage
- setup SonarCloud for Pull Requests

@JC5

### CODE REVIEW

**Comment by JC5** [comment]
Nice work, let me know what you need from me :+1:

**Comment by tonicospinelli** [comment]
@JC5 I found weird behavior using `$skip` parameter of `addPeriod` method from the `Navigation` class. Help me to understand what is expected. Considering the base date is **2019-01-31** in a **daily frequency** adding **10 days** to skip, **it expects** the calculated date to be **2019-02-10**, but the result is 11 days after returning the date _2019-02-11_.

**Comment by JC5** [comment]
Yes, that is correct. In the same sense, setting "skip" to 1 will give you a date that is 2 days later.

File: tests/Support/NavigationTest.php [hunk 1]
```diff
@@ -94,7 +96,7 @@ public static function providePeriodsWithSkippingParam(): array
             '2023-05-31 to 2023-07-31' => ['skip' => 1, 'frequency' => 'monthly', 'from' => Carbon::parse('2023-05-31'), 'expected' => Carbon::parse('2023-07-31')],
             '2023-08-31 to 2023-10-31' => ['skip' => 1, 'frequency' => 'monthly', 'from' => Carbon::parse('2023-08-31'), 'expected' => Carbon::parse('2023-10-31')],
             '2023-10-31 to 2023-12-31' => ['skip' => 1, 'frequency' => 'monthly', 'from' => Carbon::parse('2023-10-31'), 'expected' => Carbon::parse('2023-12-31')],
-            '2023-01-31 to 2023-03-30' => ['skip' => 2, 'frequency' => 'monthly', 'from' => Carbon::parse('2023-01-31'), 'expected' => Carbon::parse('2023-05-1')],
+            '2023-01-31 to 2023-03-30' => ['skip' => 2, 'frequency' => 'monthly', 'from' => Carbon::parse('2023-01-31'), 'expected' => Carbon::parse('2023-04-30')],
```
Comment (tonicospinelli) [inline tests/Support/NavigationTest.php:99]: @JC5, see here. I found a bug in the original implementation. It was disregarded in the subtraction if the date difference was more than two months.

Comment (tonicospinelli) [inline tests/Support/NavigationTest.php:99]: @JC5, see [here](https://github.com/firefly-iii/firefly-iii/pull/7685/commits/ac53eb7cd784f722714788f845a51be24fb10e94#diff-a561aad172d8a865422ff210ed6b0b58b3cc11f4b6e4c4458c5c4afa54ca1949L99-L110). I found a bug in the original implementation. It was disregarded in the subtraction if the date difference was more than two months.

**Comment by tonicospinelli** [comment]
@JC5, the changes are done 🎉

**Comment by JC5** [comment]
The variable is right there, but you may not be authorized to run it? I'm not sure. Either way, I've spotted a few small things after which i can merge. Great work!

File: app/Support/Calendar/Calculator.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
+<?php
+
+/**
+ * Copyright (c) 2023 james@firefly-iii.org
```
Comment (JC5) [inline app/Support/Calendar/Calculator.php:4]: Feel free to replace this with your own name and link to GitHub. As the author, you have the copyright, even when it's released under the AGPLv3.

File: app/Support/Calendar/Exceptions/IntervalException.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+use JetBrains\PhpStorm\Pure;
+
+final class IntervalException extends \Exception
+{
+    protected $message = 'The periodicity %s is unknown. Choose one of available periodicity: %s';
+
+    public readonly Periodicity $periodicity;
+    public readonly array $availableIntervals;
+
+    public static function unavailable(Periodicity $periodicity, array $instervals, int $code = 0, ?\Throwable $previous = null): IntervalException
```
Comment (JC5) [inline app/Support/Calendar/Exceptions/IntervalException.php:15]: Bad spelling in variable

File: app/Support/Calendar/Exceptions/IntervalException.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace FireflyIII\Support\Calendar\Exceptions;
+
+use FireflyIII\Support\Calendar\Periodicity;
+use JetBrains\PhpStorm\Pure;
```
Comment (JC5) [inline app/Support/Calendar/Exceptions/IntervalException.php:6]: Feel free to drop the Jetbrains specific class things.

**Comment by tonicospinelli** [comment]
@JC5 I made the changes that you spotted 🤗 but the sonarcloud still be broken

**Pull request merged**